docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md

```diff
@@ -0,0 +1,212 @@
+7. `gardenctl`
+    - initiates `ssh` session once `status.conditions['BastionReady']` is true of the `Bastion` resource in the garden cluster
+        - locates private `ssh` key matching `spec["sshPublicKey"]` which was configured beforehand by the user
+        - reads bastion IP (`status.ingress.ip`) or hostname (`status.ingress.hostname`)
+        - reads the private key from the `ssh` key pair for the shoot node
+        - opens `ssh` connection to the bastion and from there to the respective shoot node
+    - runs heartbeat in parallel as long as the `ssh` session is open by annotating the `Bastion` resource with `operations.gardener.cloud/operation: keepalive`
+8. `GCM`:
+    - Once `status.expirationTimestamp` is reached, the `Bastion` will be marked for deletion
+9. `gardenlet`:
+    - Once the `Bastion` resource in the garden cluster is marked for deletion, it marks the `Bastion` resource in the seed for deletion
+10. Gardener extension provider <infra> / Bastion Controller on Seed:
+    - all created resources will be cleaned up
+    - On succes, removes finalizer on `Bastion` resource in seed
+11. `gardenlet`:
+    - removes finalizer on `Bastion` resource in garden cluster
+
+### Resource Example
+
+`Bastion` resource in the garden cluster
+```yaml
+apiVersion: operations.gardener.cloud/v1alpha1
+kind: Bastion
+metadata:
+  generateName: cli-
+  name: cli-abcdef
+  namespace: garden-myproject
+  annotations:
+    operations.gardener.cloud/created-by: foo # set by the mutating webhook
+    # operations.gardener.cloud/operation: keepalive # this annotation is removed by the GAPI and the last-heartbeat timestamp and the status.expirationTimestamp will be updated accordingly
+spec:
+  shootRef: # namespace cannot be set / it's the same as .metadata.namespace
+    name: my-cluster
+
+  sshPublicKey: c3NoLXJzYSAuLi4K # public `ssh` key of the user
+
+  ingress:
+  - ipBlock:
+      cidr: 1.2.3.4/32 # public IP of the user. CIDR is a string representing the IP Block. Valid examples are "192.168.1.1/24" or "2001:db9::/64"
+
+status:
+  # the following fields are set by the GCM
+  seedName: aws-eu2
+  providerType: aws
+
+  # the following fields are managed by the controller in the seed and synced by gardenlet
+  ingress: # IP or hostname of the bastion
+    ip: 1.2.3.5
+    # hostname: foo.bar
+  conditions:
+  - type: BastionReady # when the `status` is true of condition type `BastionReady`, the client can initiate the `ssh` connection
+    status: 'True'
+    lastTransitionTime: "2021-03-19T11:59:00Z"
+    lastUpdateTime: "2021-03-19T11:59:00Z"
+    reason: BastionReady
+    message: Bastion for the cluster is ready.
+
+  # the following fields are only set by the GAPI
+  lastHeartbeatTimestamp: "2021-03-19T11:58:00Z" # will be set when setting the annotation operations.gardener.cloud/operation: keepalive
```

Comment:
Can we use `Leases` for this?

Reply:
see discussion thread here where the same question was asked https://github.com/gardener/gardener/pull/3802#discussion_r614625200